### DISCUSSION TITLE
Create 2D views with dimensions, sections etc for a production drawing

### DISCUSSION THREAD

**Discussion by Jopie01** [CadQuery/cadquery#1658] [Ideas]
As the title describes it, this is not about creating production drawings but create a 2D view with dimensions and other annotations. That view then can be exported as a vector image and then it's up to the user to put several views together with a template to create the production drawing.

So the idea is to create a `Plane` where the dimensions and annotations live. The model can also be clipped by that plane and hatched. This all should also work with assemblies. Adding dimensions can be done by selecting edges or vertices and maybe even create points in space to connect a dimension to.

There is an example how to create dimensions which comes with OpenCascade in `/usr/share/opencascade/samples/tcl/dimensions.tcl` using TCL. It's still a 3D view, but I hope it can somehow be converted into a SVG image or something like that.

I looked into https://dev.opencascade.org/doc/refman/html/package_xcafdimtolobjects.html but there I got lost and I have no idea to go any further. If anyone can shed any light on how to proceed would be awesome!

I'm fully aware that automatically adding dimensions can be problematic, but I think for simple use-cases this can be very useful. Like simple 3D box like models and simple assemblies like steel or wooden frames.

Any other thoughts on this?

**Comment by adam-urbanczyk** [comment]
I think your best bet is extending DXF export and adding dimensions via ezdxf.

**Comment by Jopie01** [comment]
> I think your best bet is extending DXF export and adding dimensions via ezdxf.

I would rather be staying with CadQuery to add the dimensions so only one interface for all the work, even ezdxf is used for importing DXF files. If using native OpenCascade is hard or troublesome maybe something like build123d can be done? Also cq_warehouse has something like that. Probably that's the way for know I think.

I hoped native OpenCascade dimensions was something which could be added to CadQuery. Currently it's not my usecase but STEP ap214 which is supported by OpenCascade can also export the annotations like dimensions etc. So when you open that STEP file in another program that supports ap214, the dimensions etc are also shown.

**Reply by adam-urbanczyk** [reply]
CadQuery as such does not provide dims. Both OCP and EZDXF are its deps, so I'm not sure what you want. Based on the original post I though 2D drawings with dims, but I'm not sure anymore. OCCT XCAF supports dims and it should be possible to render them and maybe even export to STEP, but transferring this to 2D might be tricky. Anyway, if you want to work on adding 3D dims that would be welcome in the assy layer. If you want to work on 2D, I'd propose to start here: https://github.com/CadQuery/cadquery/blob/6d7d1d640d87410913f81b96babf914fd45f77a3/cadquery/occ_impl/exporters/dxf.py#L45

Also take a look at: https://github.com/CadQuery/cadquery/issues/122

**Reply by Jopie01** [reply]
> CadQuery as such does not provide dims. Both OCP and EZDXF are its deps, so I'm not sure what you want. Based on the original post I though 2D drawings with dims

Let first clear up what '2D drawings' is. A 2D drawing:
1. is a template with information about the product on the drawing  
2. has several views (left, front, top, iso etc)
3. each view can have several dimensions or other annotations

In my start post I talked about point 3. And because adding those dimensions in 3D through OpenCascade should make it also possible to export them to a STEP ap214 file. But DXF is indeed another thing.

In several 3D CAD programs I'm able to create a plane and use that plane as the annotation plane. Dimensions etc are placed on that plane. It seems OpenCascade can do that as well.

> Anyway, if you want to work on adding 3D dims that would be welcome in the assy layer.

I will try to see how far I can get with this. There is not a lot of documentation how to add dimensions. But am I right  that exporting to DXF has nothing to do with OpenCascade?

In my previous post I missed the link to the build123d drafting file https://github.com/gumyr/build123d/blob/dev/src/build123d/drafting.py . Maybe that's an easier way.

**Comment by Jopie01** [comment]
In the end, it was indeed the best bet to use `ezdxf` to create and export a DXF file. I'm still at the start of this journey but `ezdxf` is much, much more capable then I thought. It also meant that I had to create a complete new DXF export system because the current one lacks the possibilities that I need. I want a multi-page DXF which means using 'paperspace' for the different layouts and add viewports, dimensions and notes to it. Also multiple different views should be added like front, left and top.

So the steps I'm taking now is:
1. open a DXF template with the right dimensionstyles, layers etc
3. create a view based on the angles send by the user and put that on a layer in modelspace
4. create a paperspace layout and add the drawing template with information
5. create a viewport to show the view in modelspace
6. do 2 and 4 again for more views
7. eventually add more layouts, more views, more viewports
8. add dimensions and notes to a layout
9. save the changes to a new DXF file

**Reply by adam-urbanczyk** [reply]
Thanks for sharing! Is this something you'd be willing to contribute?

**Reply by Jopie01** [reply]
I think I cannot contribute everything back because several parts like dimensions and notes are very specific for my customer. But to make DXF export a lot better is to use the same technique as with SVG; using `HLRBRep_Algo` and `HLRAlgo_Projector` to get the lines. This allows for hidden lines etc which is now impossible in DXF export.

Also implementing issue #1685 for the `projectionDir` made it possible to get the right view easy (maybe it also helps to get the coordinate system drawn correctly).

I haven't looked too deeply into the current DXF export, but I think you already can add layers and different shapes.
